packages/react/src/dialog/popup/DialogPopup.tsx

```diff
@@ -39,44 +39,45 @@ export const DialogPopup = React.forwardRef(function DialogPopup(
 ) {
   const { className, finalFocus, initialFocus, render, ...elementProps } = componentProps;
 
-  const {
-    descriptionElementId,
-    dismissible,
-    floatingRootContext,
-    getPopupProps,
-    modal,
-    mounted,
-    nested,
-    nestedOpenDialogCount,
-    setOpen,
-    open,
-    openMethod,
-    popupRef,
-    setPopupElement,
-    titleElementId,
-    transitionStatus,
-    onOpenChangeComplete,
-    internalBackdropRef,
-  } = useDialogRootContext();
+  const { store } = useDialogRootContext();
+
+  const descriptionElementId = store.useState('descriptionElementId');
+  const dismissible = store.useState('dismissible');
+  const floatingRootContext = store.useState('floatingRootContext');
+  const rootPopupProps = store.useState('popupProps');
+  const modal = store.useState('modal');
+  const mounted = store.useState('mounted');
+  const nested = store.useState('nested');
+  const nestedOpenDialogCount = store.useState('nestedOpenDialogCount');
+  const open = store.useState('open');
+  const openMethod = store.useState('openMethod');
+  const titleElementId = store.useState('titleElementId');
+  const transitionStatus = store.useState('transitionStatus');
 
   useDialogPortalContext();
 
   useOpenChangeComplete({
     open,
-    ref: popupRef,
+    ref: store.elements.popupRef,
     onComplete() {
       if (open) {
-        onOpenChangeComplete?.(true);
+        store.events.emit('openChangeComplete', true);
```

Comment:
Slightly confused about the `events` part. Why is this used instead of `store.state.onOpenChangeComplete?.(true)`

Reply:
Alternatively, we could have `store.eventHandlers.onOpenChangeComplete` registered with a custom hook that wraps the provided prop in `useEventCallback` - this would also separate these handlers from state and prevent unnecessary updates. Plus, remove this `events.emit` indirection.

Reply:
I'm not really a fan of having random new properties that increases the API surface to understand/learn

I'd still consider it "state" in the sense that it's derived from props and represents a given value at a given time, even if it lacks an associated setter